### ISSUE TITLE
lint unnecessary derefs

### ISSUE THREAD

**Issue by birkenfeld** [Manishearth/rust-clippy#234]
I've seen uses of `&*s` where `s` is a `String`; this can be reduced to `&s` when calling a function that accepts `&str`.

Similarly, in the AST, I've seen (in clippy and racer) `&**x` or `&*x` where just `x` or `&x` is fine; call syntax will autoderef a `&P<T>` to a `&T`.

However, I don't really know yet how to do this in a general way.

**Comment by llogiq** [comment]
I do have some instances of this in `shadow.rs`; however, I only added those because `rustc` would not compile it otherwise. Even knowing the general rules of auto-deref, I am not convinced I understand all the strange corner cases.

So unless we use a blacklist (same old, same old :smile:), we'll have a real hard time trying to avert false positives.

**Comment by birkenfeld** [comment]
Yeah, as far as I understand it, the adjustment is done for function arguments and method/field lookup. But inner types are not adjusted, so once you have `Option<P<...>>` you need to do the dance manually.

This would have to check function arguments against the parameter types, and we'll have to find if rustc provides an API that does the "is type A compatible with type B after adjusting" check. If it doesn't, this is going to be hard :)

**Comment by llogiq** [comment]
Agreed, you either end up building your own types at runtime (hard) or rewriting most of the type adjustment logic (probably harder).

**Comment by fpoli** [comment]
Even just supporting the case `&*s` where `s` is a `String` would be very useful. New Rust users usually struggle a lot understanding the difference between `str`, `&str`, `String` etc. It's probably one of the most common types with auto deref.

**Comment by lengyijun** [comment]
I wonder if `&*String` => `&String` is really good?
`&*String` always return `&str`
`&String` return `&str` or `&String`

**Comment by fpoli** [comment]
> I wonder if `&*String` => `&String` is really good?
> `&*String` always return `&str`
> `&String` return `&str` or `&String`

That's a good point. The lint could be reported only if `&*String` expression is in the place of a method argument that requires `&str`.

**Comment by lengyijun** [comment]
```
let mut v=vec![];
// they are different!
v.push(s);
v.push(&**s); 
```
